### ISSUE TITLE
Wrong inference in a metaclass when `self` is annotated as `Type[T]` in `__get__`'s overload

### ISSUE THREAD

**Issue by jack1142** [microsoft/pyright#2177]
**Describe the bug**
Annotating `self` in a metaclass as `Type[T]` breaks inference for the descriptor.

**To Reproduce**
1. Copy code from below
2. Run it, you should see:
```
<class '__main__.IntDescriptorClass'>
<__main__.IntDescriptorClass object at 0x00000243AB1A7880>
```
3. Type check it with pyright, as part of the output you should see:
```
metaclass_descriptors.py:31:17 - info: Type of "X.number_cls" is "Type[IntDescriptorClass]"
metaclass_descriptors.py:32:17 - info: Type of "X().number_cls" is "Type[IntDescriptorClass]"
```

**Expected behavior**
I expected pyright to infer type of `X.number_cls`/`X().number_cls` correctly:
```
metaclass_descriptors.py:31:17 - info: Type of "X.number_cls" is "Type[IntDescriptorClass]"
metaclass_descriptors.py:32:17 - info: Type of "X().number_cls" is "IntDescriptorClass"
```

**Screenshots or Code**
```python
from typing import Any, Optional, Union, Type, TypeVar, TYPE_CHECKING, overload


T = TypeVar("T")


class _IntDescriptorMeta(type):
    @overload
    def __get__(self: Type[T], instance: None, owner: Any) -> Type[T]: ...

    @overload
    def __get__(self: Type[T], instance: object, owner: Any) -> T: ...

    def __get__(self: Type[T], instance: Optional[object], owner: Any) -> Union[Type[T], T]:
        if instance is None:
            return self
        return self()

class IntDescriptorClass(metaclass=_IntDescriptorMeta):
    ...


class X:
    number_cls = IntDescriptorClass


print(X.number_cls)
print(X().number_cls)

if TYPE_CHECKING:
    reveal_type(X.number_cls)
    reveal_type(X().number_cls)
```

**VS Code extension or command-line**
Running pyright 1.1.161 as a command-line tool. Replicable on the build from the current state of `main` branch as well.

**Additional context**
N/A

**Comment by erictraut** [comment]
Thanks for the bug report.

FWIW, mypy gets this wrong as well.

**Comment by jack1142** [comment]
> FWIW, mypy gets this wrong as well.

I made an issue for #2164 yesterday on mypy's repo yesterday so I'll comment on it with a link to this one as well to make sure this is supported properly if/when they add support for descriptor metaclasses, thanks!

**Comment by erictraut** [comment]
This will be implemented in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.162, which I just published. It will also be in the next release of pylance.
